src/util/browser.js

```diff
@@ -37,15 +39,22 @@ const exported = {
     },
 
     getImageData(img: CanvasImageSource, padding?: number = 0): ImageData {
-        const canvas = window.document.createElement('canvas');
+        const {width, height} = img;
+
+        if (!canvas) {
+            canvas = window.document.createElement('canvas');
+        }
+
         const context = canvas.getContext('2d');
         if (!context) {
             throw new Error('failed to create canvas 2d context');
         }
-        canvas.width = img.width;
-        canvas.height = img.height;
-        context.drawImage(img, 0, 0, img.width, img.height);
-        return context.getImageData(-padding, -padding, img.width + 2 * padding, img.height + 2 * padding);
+
+        canvas.width = width > canvas.width ? width : canvas.width;
+        canvas.height = height > canvas.height ? height : canvas.height;
+
+        context.drawImage(img, 0, 0, width, height);
+        return context.getImageData(-padding, -padding, width + 2 * padding, height + 2 * padding);
```

Comment:
Following up from comment https://github.com/mapbox/mapbox-gl-js/pull/11564#discussion_r819983660, I think that would be the right way to resize:
```suggestion
        if (width > canvas.width || height > canvas.height) {
            canvas.width = width;
            canvas.height = height;
        }

        context.clearRect(-padding, -padding, width + 2 * padding, height + 2 * padding);
        context.drawImage(img, 0, 0, width, height);
        return context.getImageData(-padding, -padding, width + 2 * padding, height + 2 * padding);
```
The canvas might be dropped as soon as we resize it, to retain it we still need the condition but have a clear on the region where we'd like to use it. The above suggestion fixes that issue with proper caching.